File: machine-configuration/desktop/appearance/screensaver/ambient-canvas/scripts/ambient_canvas_media/ambient_canvas_shuffled_segment_order.py
import random


class ShuffledSegmentOrder:
    def __init__(self, segment_sequence_identifiers, random_source=None):
        self.random_source = random_source or random
        self.segment_indices_by_choice = []
        choice_index_by_sequence = {}
        for segment_index, sequence_identifier in enumerate(
            segment_sequence_identifiers
        ):
            if sequence_identifier:
                choice_index = choice_index_by_sequence.get(sequence_identifier)
                if choice_index is None:
                    choice_index = len(self.segment_indices_by_choice)
                    choice_index_by_sequence[sequence_identifier] = choice_index
                    self.segment_indices_by_choice.append([])
            else:
                choice_index = len(self.segment_indices_by_choice)
                self.segment_indices_by_choice.append([])
            self.segment_indices_by_choice[choice_index].append(segment_index)
        self.next_segment_position_by_choice = [
            0 for _ in self.segment_indices_by_choice
        ]
        self.remaining_choice_indices = []
        self.previously_played_choice_index = None

    def next_segment_index(self):
        if not self.remaining_choice_indices:
            self.refill_avoiding_immediate_repeat()
        chosen_choice_index = self.remaining_choice_indices.pop(0)
        segment_indices = self.segment_indices_by_choice[chosen_choice_index]
        segment_position = self.next_segment_position_by_choice[chosen_choice_index]
        chosen_segment_index = segment_indices[segment_position]
        self.next_segment_position_by_choice[chosen_choice_index] = (
            segment_position + 1
        ) % len(segment_indices)
        self.previously_played_choice_index = chosen_choice_index
        return chosen_segment_index

    def refill_avoiding_immediate_repeat(self):
        choice_count = len(self.segment_indices_by_choice)
        if choice_count <= 0:
            return
        shuffled_indices = list(range(choice_count))
        self.random_source.shuffle(shuffled_indices)
        if (
            choice_count > 1
            and shuffled_indices[0] == self.previously_played_choice_index
        ):
            shuffled_indices[0], shuffled_indices[-1] = (
                shuffled_indices[-1],
                shuffled_indices[0],
            )
        self.remaining_choice_indices = shuffled_indices

File: machine-configuration/desktop/appearance/screensaver/ambient-canvas/swift-sources/ambient-canvas-shuffled-segment-order.swift
import Foundation

final class AmbientCanvasShuffledSegmentOrder {
    private let segmentIndicesByChoice: [[Int]]
    private var nextSegmentPositionByChoice: [Int]
    private var remainingChoiceIndices: [Int] = []
    private var previouslyPlayedChoiceIndex: Int?

    init(sequenceIdentifiers: [String?]) {
        var resolvedSegmentIndicesByChoice: [[Int]] = []
        var choiceIndexBySequence: [String: Int] = [:]
        for (segmentIndex, sequenceIdentifier) in sequenceIdentifiers.enumerated() {
            let choiceIndex: Int
            if let sequenceIdentifier, !sequenceIdentifier.isEmpty {
                if let existingChoiceIndex = choiceIndexBySequence[sequenceIdentifier] {
                    choiceIndex = existingChoiceIndex
                } else {
                    choiceIndex = resolvedSegmentIndicesByChoice.count
                    choiceIndexBySequence[sequenceIdentifier] = choiceIndex
                    resolvedSegmentIndicesByChoice.append([])
                }
            } else {
                choiceIndex = resolvedSegmentIndicesByChoice.count
                resolvedSegmentIndicesByChoice.append([])
            }
            resolvedSegmentIndicesByChoice[choiceIndex].append(segmentIndex)
        }
        segmentIndicesByChoice = resolvedSegmentIndicesByChoice
        nextSegmentPositionByChoice = Array(
            repeating: 0,
            count: resolvedSegmentIndicesByChoice.count
        )
    }

    func nextSegmentIndex() -> Int {
        if remainingChoiceIndices.isEmpty {
            refillAvoidingImmediateRepeat()
        }
        let chosenChoiceIndex = remainingChoiceIndices.removeFirst()
        let segmentIndices = segmentIndicesByChoice[chosenChoiceIndex]
        let segmentPosition = nextSegmentPositionByChoice[chosenChoiceIndex]
        let chosenSegmentIndex = segmentIndices[segmentPosition]
        nextSegmentPositionByChoice[chosenChoiceIndex] =
            (segmentPosition + 1) % segmentIndices.count
        previouslyPlayedChoiceIndex = chosenChoiceIndex
        return chosenSegmentIndex
    }

    private func refillAvoidingImmediateRepeat() {
        guard !segmentIndicesByChoice.isEmpty else {
            return
        }
        var shuffledIndices = Array(0..<segmentIndicesByChoice.count).shuffled()
        if
            segmentIndicesByChoice.count > 1,
            shuffledIndices.first == previouslyPlayedChoiceIndex
        {
            shuffledIndices.swapAt(0, shuffledIndices.count - 1)
        }
        remainingChoiceIndices = shuffledIndices
    }
}
